File: pkg/session/session.go
// Package session provides session ID derivation from git branch names
package session

import (
	"context"
	"fmt"
	"os/exec"
	"regexp"
	"strings"

	"github.com/cortex-ai/cortex-ai/internal/config"
	"github.com/google/uuid"
)

// Deriver handles session ID derivation from git branch names
type Deriver struct {
	cfg *config.SessionConfig
}

// NewDeriver creates a new session ID deriver
func NewDeriver(cfg *config.SessionConfig) *Deriver {
	return &Deriver{cfg: cfg}
}

// GetCurrentBranch returns the current git branch name
func GetCurrentBranch(ctx context.Context) (string, error) {
	cmd := exec.CommandContext(ctx, "git", "rev-parse", "--abbrev-ref", "HEAD")
	output, err := cmd.Output()
	if err != nil {
		return "", fmt.Errorf("failed to get git branch: %w", err)
	}

	branch := strings.TrimSpace(string(output))
	if branch == "" {
		return "", fmt.Errorf("empty branch name")
	}

	return branch, nil
}

// DeriveSessionID derives a session ID from the current git branch
// Returns the session ID or an error if derivation fails
func (d *Deriver) DeriveSessionID(ctx context.Context) (string, error) {
	// Get current branch
	branch, err := GetCurrentBranch(ctx)
	if err != nil {
		if d.cfg.FallbackToUUID {
			return uuid.New().String(), nil
		}
		return "", fmt.Errorf("failed to get branch name: %w", err)
	}

	return d.DeriveFromBranch(branch)
}

// DeriveFromBranch derives a session ID from a given branch name
func (d *Deriver) DeriveFromBranch(branch string) (string, error) {
	// Strip prefix if configured
	if d.cfg.StripPrefix != "" {
		branch = strings.TrimPrefix(branch, d.cfg.StripPrefix)
	}

	var sessionID string
	var err error

	switch d.cfg.PatternType {
	case "prefix":
		sessionID = d.extractPrefix(branch)
	case "regex":
		sessionID, err = d.extractRegex(branch)
		if err != nil {
			if d.cfg.FallbackToUUID {
				return uuid.New().String(), nil
			}
			return "", err
		}
	case "full":
		sessionID = d.extractFull(branch)
	default:
		// Default to prefix mode
		sessionID = d.extractPrefix(branch)
	}

	// If empty and fallback is enabled, use UUID
	if sessionID == "" && d.cfg.FallbackToUUID {
		return uuid.New().String(), nil
	}

	if sessionID == "" {
		return "", fmt.Errorf("failed to derive session ID from branch: %s", branch)
	}

	return sessionID, nil
}

// extractPrefix extracts a prefix-based session ID
// Example: "fix/sil-123/do-implementation" -> "session-fix-sil-123"
func (d *Deriver) extractPrefix(branch string) string {
	// Replace slashes with configured separator
	parts := strings.Split(branch, "/")

	// Apply max segments limit
	if d.cfg.MaxSegments > 0 && len(parts) > d.cfg.MaxSegments {
		parts = parts[:d.cfg.MaxSegments]
	}

	// Join with separator
	sessionPart := strings.Join(parts, d.cfg.Separator)

	// Add prefix
	return d.cfg.Prefix + sessionPart
}

// extractRegex extracts session ID using a custom regex pattern
func (d *Deriver) extractRegex(branch string) (string, error) {
	if d.cfg.Pattern == "" {
		return "", fmt.Errorf("regex pattern is empty")
	}

	re, err := regexp.Compile(d.cfg.Pattern)
	if err != nil {
		return "", fmt.Errorf("invalid regex pattern: %w", err)
	}

	matches := re.FindStringSubmatch(branch)
	if len(matches) < 2 {
		return "", fmt.Errorf("regex pattern did not match branch: %s", branch)
	}

	// Use first capture group
	sessionPart := matches[1]

	// Replace slashes with separator
	sessionPart = strings.ReplaceAll(sessionPart, "/", d.cfg.Separator)

	// Add prefix
	return d.cfg.Prefix + sessionPart, nil
}

// extractFull uses the full branch name as session ID
func (d *Deriver) extractFull(branch string) string {
	// Replace slashes with separator
	sessionPart := strings.ReplaceAll(branch, "/", d.cfg.Separator)

	// Add prefix
	return d.cfg.Prefix + sessionPart
}

// DeriveOrUseProvided returns the provided session ID if not empty,
// otherwise derives it from the git branch
func (d *Deriver) DeriveOrUseProvided(ctx context.Context, providedSessionID string) (string, error) {
	// If session ID is explicitly provided, use it
	if providedSessionID != "" {
		return providedSessionID, nil
	}

	// If auto-derive is disabled, return empty
	if !d.cfg.AutoDerive {
		return "", nil
	}

	// Derive from git branch
	return d.DeriveSessionID(ctx)
}

File: src/cortex/session.py
"""Session ID derivation from git branch names."""

from __future__ import annotations

import re
import subprocess
from uuid import uuid4

from cortex.config.settings import SessionConfig


def derive_session_id(config: SessionConfig) -> str:
    """Derive a session ID from the current git branch.

    Falls back to a UUID if not in a git repo or if the branch
    cannot be determined and fallback_to_uuid is enabled.

    Raises RuntimeError if auto_derive fails and fallback is disabled.
    """
    if not config.auto_derive:
        return _uuid_session(config)

    branch = _get_git_branch()
    if not branch:
        if config.fallback_to_uuid:
            return _uuid_session(config)
        raise RuntimeError("Could not determine git branch for session ID derivation")

    return _branch_to_session(branch, config)


def _get_git_branch() -> str | None:
    """Return the current git branch name, or None if unavailable."""
    try:
        result = subprocess.run(
            ["git", "rev-parse", "--abbrev-ref", "HEAD"],
            capture_output=True,
            text=True,
            timeout=5,
        )
        if result.returncode == 0:
            branch = result.stdout.strip()
            return branch if branch and branch != "HEAD" else None
    except Exception:
        pass
    return None


def _branch_to_session(branch: str, config: SessionConfig) -> str:
    """Convert a git branch name to a session ID per config rules."""
    pattern_type = config.pattern_type

    if pattern_type == "full":
        slug = _slugify(branch)
        return f"{config.prefix}{slug}"

    if pattern_type == "regex" and config.pattern:
        match = re.search(config.pattern, branch)
        if match:
            slug = _slugify(match.group(0))
            return f"{config.prefix}{slug}"
        # Fall through to prefix mode on regex miss

    # Default: prefix mode — split on delimiters, take max_segments
    # Strip any configured prefix from the branch first
    working = branch
    if config.strip_prefix and working.startswith(config.strip_prefix):
        working = working[len(config.strip_prefix) :]

    segments = re.split(r"[/\-_]", working)
    segments = [s for s in segments if s][: config.max_segments]
    slug = config.separator.join(segments)
    return f"{config.prefix}{slug}"


def _uuid_session(config: SessionConfig) -> str:
    return f"{config.prefix}{uuid4()}"


def _slugify(text: str) -> str:
    """Convert text to a safe slug for session IDs."""
    return re.sub(r"[^a-zA-Z0-9\-_]", "-", text).strip("-")
